library/tiqr/Tiqr/Message/Firebase.php

```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * This file is part of the tiqr project.
+ * 
+ * The tiqr project aims to provide an open implementation for 
+ * authentication using mobile devices. It was initiated by 
+ * SURFnet and developed by Egeniq.
+ *
+ * More information: http://www.tiqr.org
+ *
+ * @author Joost van Dijk <joost.vandijk@surfnet.nl>
+ * 
+ * @package tiqr
+ *
+ * @license New BSD License - See LICENSE file for details.
+ *
+ * @copyright (C) 2010-2015 SURFnet BV
+ */
+
+
+/** @internal base includes */
+require_once('Tiqr/Message/Abstract.php');
+
+/**
+ * Android Cloud To Device Messaging message.
+ *
+ * This implementation is used as a fallback GCM send event.
+ */
+class Tiqr_Message_Firebase extends Tiqr_Message_Abstract
+{
+    /**
+     * Send message.
+     *
+     * @throws Tiqr_Message_Exception_AuthFailure
+     * @throws Tiqr_Message_Exception_SendFailure
+     */
+    public function send()
+    {
+        $options = $this->getOptions();
+        $apiKey      = $options['firebase.apikey'];
+
+        $translatedAddress = $this->getAddress();
+        $name = $this->getText();
+        $url = $this->getCustomProperty('challenge');
+        $this->_push($translatedAddress, "Please authenticate for " . $name, $url, $apiKey);
+    }
+
+    /**
+     * Send a message to a device using the firebase API key.
+     *
+     * @param $deviceToken string device ID
+     * @param $alert string alert message
+     * @param $challenge string tiqr challenge url
+     * @param $apiKey string api key for firebase
+     *
+     * @throws Tiqr_Message_Exception_SendFailure
+     */
+    private function _push($deviceToken, $alert, $challenge, $apiKey)
```

Comment:
This is not your design decision, I'm aware of that. But a message that is also able to send itself is not very solid. 

/observation.